crates/hotshot/src/traits/networking/push_cdn_network.rs

```diff
@@ -0,0 +1,424 @@
+    /// always.
+    async fn is_ready(&self) -> bool {
+        true
+    }
+
+    /// TODO: shut down the networks. Unneeded for testing.
+    fn shut_down<'a, 'b>(&'a self) -> BoxSyncFuture<'b, ()>
+    where
+        'a: 'b,
+        Self: 'b,
+    {
+        // TODO: figure out if drop is sufficient
+        boxed_sync(async move {})
+    }
+
+    /// Broadcast a message to all members of the quorum.
+    ///
+    /// # Errors
+    /// - If we fail to serialize the message
+    /// - If we fail to send the broadcast message.
+    async fn broadcast_message(
+        &self,
+        message: Message<TYPES>,
+        _recipients: BTreeSet<TYPES::SignatureKey>,
+    ) -> Result<(), NetworkError> {
+        self.broadcast_message(message, Topic::Global).await
+    }
+
+    /// Broadcast a message to all members of the DA committee.
+    ///
+    /// # Errors
+    /// - If we fail to serialize the message
+    /// - If we fail to send the broadcast message.
+    async fn da_broadcast_message(
+        &self,
+        message: Message<TYPES>,
+        _recipients: BTreeSet<TYPES::SignatureKey>,
+    ) -> Result<(), NetworkError> {
+        self.broadcast_message(message, Topic::DA).await
+    }
+
+    /// Send a direct message to a node with a particular key. Does not retry.
+    ///
+    /// - If we fail to serialize the message
+    /// - If we fail to send the direct message
+    async fn direct_message(
+        &self,
+        message: Message<TYPES>,
+        recipient: TYPES::SignatureKey,
+    ) -> Result<(), NetworkError> {
+        // Bincode the message
+        let serialized_message = match bincode_opts().serialize(&message) {
+            Ok(serialized) => serialized,
+            Err(e) => {
+                warn!("Failed to serialize message: {}", e);
+                return Err(NetworkError::FailedToSerialize { source: e });
+            }
+        };
+
+        // Send the message
+        // TODO: check if we need to print this error
+        if self
+            .0
+            .send_direct_message(&WrappedSignatureKey(recipient), serialized_message)
+            .await
+            .is_err()
+        {
+            return Err(NetworkError::CouldNotDeliver);
+        };
+
+        Ok(())
+    }
+
+    /// Receive a message. Is agnostic over `transmit_type`, which has an issue
+    /// to be removed anyway.
+    ///
+    /// # Errors
+    /// - If we fail to receive messages. Will trigger a retry automatically.
+    async fn recv_msgs(&self) -> Result<Vec<Message<TYPES>>, NetworkError> {
+        // Receive a message
+        let message = self.0.receive_message().await;
+
+        // If it was an error, wait a bit and retry
+        let message = match message {
+            Ok(message) => message,
+            Err(error) => {
+                async_sleep(Duration::from_millis(100)).await;
+                error!("failed to receive message: {error}");
```

Comment:
What's the reason for sleeping after an error?  I tried to dig into the CDN code but didn't really find any cases where sleeping would fix an error case, but I very easily could be missing something.

Reply:
The initial reason I used a sleep here is because a failed message will always lead to a a reconnect under the hood, which will take some time. I wanted to avoid spinning cycles if, for some reason, we were failing the connection immediately. 

But removing this, as I just checked and we have a sleep in the actual client retry code.